### PR TITLE
Support multiple forks in the bank

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Codebase isn't designed around a bank that has multiple forks.

#### Summary of Changes

Added the the constraint that bank state can be forked. Propagating all the changes through the rest of the system.  Factored out components from the bank into their own modules. (Quit adding things to bank that can be in its own file!!!)

This can be merged separately
* split up duplicate signature detection from last id tracking
* use a bloom filter for duplicate signature detection

Fixes #

### CODE REVIEW

File: src/bank.rs [hunk 1]
```diff
@@ -348,333 +285,65 @@ impl Bank {
...
+        let rvs = state.par_process_entries(&states, entries);
+        //TODO: this should only be called at the end of a block by replay stage
+        self.leader_scheduler
+            .write()
+            .unwrap()
+            .update_height(self.tick_height(), self);
+        rvs
     }
 
+    /// Process an ordered list of entries, populating a circular buffer "tail"
```
Comment (rob-solana) [inline src/bank.rs]: "tail" comments are obsolete

File: src/bloom.rs [hunk 2]
```diff
@@ -56,10 +63,51 @@ impl<T: BloomHashIndex> Bloom<T> {
     }
 }
 
+fn to_slice(v: u64) -> [u8; 8] {
+    [
+        (v & 0xffu64) as u8,
+        (v >> 1 & 0xffu64) as u8,
```
Comment (rob-solana) [inline src/bloom.rs]: should these be shifts by mulitples of 8?

File: src/checkpoints.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+impl<T> Default for Checkpoints<T> {
+    fn default() -> Self {
+        Self {
+            checkpoints: HashMap::new(),
+            latest: HashSet::new(),
+        }
+    }
+}
```
Comment (rob-solana) [inline src/checkpoints.rs]: tests for this module would help me read it

File: src/entry_queue.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+/// The number of most recent `last_id` values that the bank will track the signatures
+/// of. Once the bank discards a `last_id`, it will reject any transactions that use
+/// that `last_id` in a transaction. Lowering this value reduces memory consumption,
+/// but requires clients to update its `last_id` more frequently. Raising the value
+/// lengthens the time a client must wait to be certain a missing transaction will
+/// not be processed by the network.
+pub const MAX_ENTRY_IDS: usize = NUM_TICKS_PER_SECOND * 120;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Entry {
```
Comment (rob-solana) [inline src/entry_queue.rs]: this looks like locktower coming in?  can we use a different name for these?

File: src/replay_stage.rs [hunk 5]
```diff
@@ -84,103 +106,62 @@ impl ReplayStage {
...
-
-        let entries_len = entries.len() as u64;
-        // TODO: In line with previous behavior, this will write all the entries even if
-        // an error occurred processing one of the entries (causing the rest of the entries to
-        // not be processed).
-        if entries_len != 0 {
-            ledger_entry_sender.send(entries)?;
-        }
+        let now = Instant::now();
+        // TODO: make this work
```
Comment (aeyakovenko) [inline src/replay_stage.rs]: @rob-solana @carllin does this TOD make sense?

File: src/entry_queue.rs (hunk 4 above)
Comment (aeyakovenko) [reply src/entry_queue.rs]: How about LastId queue?

File: src/bank_state.rs [hunk 6]
```diff
@@ -524,7 +454,7 @@ impl BankState {
 
         let now = Instant::now();
         // Once the accounts are new transactions can enter the pipeline to process them
-        self.unlock_accounts(&txs, &results);
+        head.unlock_accounts(&txs, &results);
```
Comment (rob-solana) [inline src/bank_state.rs]: this code needs a little rework, right?  record() above can fail and we're working on a fork inside the single bank per fullnode?  what happens to a fork that's done with a block (when record() fails)?

File: src/replay_stage.rs (hunk 5 above)
Comment (rob-solana) [reply src/replay_stage.rs]: yep

File: src/forks.rs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        if let Some(state) = self.checkpoints.load(base) {
+            if !state.0.finalized() {
+                return Err(BankError::CheckpointNotFinalized);
+            }
+            let new = state.0.fork(current, last_id);
+            self.checkpoints.store(current, Arc::new(new), base);
+            self.live_fork = current;
+            Ok(())
+        } else {
+            return Err(BankError::UnknownFork);
```
Comment (garious) [inline src/forks.rs]: Typical style is to put this branch first and then not intent the success case.

**Comment by garious** [comment]
@aeyakovenko, can you add issues for the various problems this PR addresses? It's not looking like this should all go in at once and I don't want to lose any of this otherwise great work.

File: src/bank.rs [hunk 8]
```diff
@@ -3,21 +3,18 @@
...
 
-use crate::accounts::{Accounts, ErrorCounters, InstructionAccounts, InstructionLoaders};
-use crate::checkpoint::Checkpoint;
-use crate::counter::Counter;
-use crate::entry::Entry;
-use crate::entry::EntrySlice;
+use crate::bank_state::{BankCheckpoint, BankState};
+use crate::entry::{Entry, EntrySlice};
+use crate::forks::Forks;
+//use crate::jsonrpc_macros::pubsub::Sink;
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/bank.rs]: @mvines can you take a look at how to re-add the pubsub stuff?

File: src/bank.rs [hunk 9]
```diff
@@ -102,53 +103,48 @@ impl BankSubscriptions for LocalSubscriptions {
...
-    /// Create an Bank with built-in programs.
-    pub fn new_with_builtin_programs() -> Self {
-        let bank = Self::default();
-        bank.add_builtin_programs();
-        bank
-    }
-
     /// Create an Bank using a deposit.
-    pub fn new_from_deposits(deposits: &[Payment]) -> Self {
+    fn new_from_deposits(deposits: &[Payment], trunk: u64, last_id: &Hash) -> Self {
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/bank.rs]: @garious root_slot_height?

File: Cargo.toml [hunk 10]
```diff
@@ -29,11 +29,11 @@ bs58 = "0.2.0"
 bv = { version = "0.11.0", features = ["serde"] }
 byteorder = "1.3.1"
 chrono = { version = "0.4.0", features = ["serde"] }
-fnv = "1.0.6"
 hashbrown = "0.1.8"
 indexmap = "1.0"
 itertools = "0.8.0"
 libc = "0.2.48"
+fnv = "1.0.6"
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply Cargo.toml]: Bummer cargo fmt doesn't do that :(

File: src/bank.rs [hunk 11]
```diff
@@ -386,362 +361,47 @@ impl Bank {
...
-            );
-        }
-        if 0 != error_counters.insufficient_funds {
-            inc_new_counter_info!(
-                "bank-process_transactions-error-insufficient_funds",
-                error_counters.insufficient_funds
-            );
-        }
-        (loaded_accounts, executed)
+    pub fn trunk_fork(&self) -> BankState {
```
[the remark this answers is not in the document]
Comment (aeyakovenko) [reply src/bank.rs]: I like root

File: src/forks.rs (hunk 7 above)
Comment (aeyakovenko) [reply src/forks.rs]: @garious there is no clean `!let Some(...) =`

**Comment by aeyakovenko** [comment]
@sakridge if i try to enable the storage test, i get this error:
```
---- bank::tests::test_bank_storage stdout ----
thread 'bank::tests::test_bank_storage' panicked at 'asser
tion failed: `(left == right)`
  left: `Err(ProgramError(0, GenericError))`,
 right: `Ok(())`', src/bank.rs:1606:10
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

File: src/accounts.rs [hunk 12]
```diff
@@ -87,22 +78,21 @@ impl AccountsDB {
...
+        for db in checkpoints {
+            if let Some(account) = db.accounts.get(pubkey) {
+                return Some(account.clone());
             }
         }
         None
     }
-
-    pub fn store(&mut self, pubkey: &Pubkey, account: &Account) {
+    /// purge == checkpoints.is_empty()
```
Comment (jackcmay) [inline src/accounts.rs]: What is this referring to?  I only see `let purge = self.checkpoints.len() == 1;` elsewhere in the code, not `is_empty()`.

File: src/accounts.rs [hunk 13]
```diff
@@ -87,22 +78,21 @@ impl AccountsDB {
...
+    pub fn store(&mut self, purge: bool, pubkey: &Pubkey, account: &Account) {
         if account.tokens == 0 {
-            if self.checkpoints.is_empty() {
+            if purge {
                 // purge if balance is 0 and no checkpoints
```
Comment (jackcmay) [inline src/accounts.rs]: This comment says to purge if the balance is zero, where is balance factored into the decision to purge?

File: src/bank.rs [hunk 14]
```diff
@@ -897,39 +595,32 @@ impl Bank {
...
 
     pub fn get_account(&self, pubkey: &Pubkey) -> Option<Account> {
-        self.accounts.load_slow(pubkey)
+        let state = self.live_bank_state();
+        state.load_slow(pubkey)
     }
 
+    pub fn store_slow(&self, pubkey: &Pubkey, account: &Account) {
+        let state = self.live_bank_state();
+        let purge = state.checkpoints.len() == 1;
```
Comment (jackcmay) [inline src/bank.rs]: Can you help me understand what the significance of `checkpoints.len == 1` and why it determines if we purge an account?

Comment (aeyakovenko) [reply src/bank.rs]: added comments

File: src/accounts.rs (hunk 13 above)
Comment (aeyakovenko) [reply src/accounts.rs]: balance check is on the line above.  I added docs to explain what purge does.

**Comment by sakridge** [comment]
@aeyakovenko `test_bank_storage` test seems to pass fine for me after uncommenting.

**Comment by aeyakovenko** [comment]
closing this because it is being merged in pieces

**Pull request closed without merging**